// @tensorflow/tfjs-layers/dist/layers/preprocessing/image_resizing.js
/**
 * @license
 * Copyright 2022 CodeSmith LLC
 *
 * Use of this source code is governed by an MIT-style
 * license that can be found in the LICENSE file or at
 * https://opensource.org/licenses/MIT.
 * =============================================================================
 */
import { image, serialization, tidy } from '@tensorflow/tfjs-core'; // mul, add
import { Layer } from '../../engine/topology';
import { ValueError } from '../../errors';
import { getExactlyOneShape } from '../../utils/types_utils'; //, getExactlyOneTensor
// tf methods unimplemented in tfjs: 'bicubic', 'area', 'lanczos3', 'lanczos5',
//                                   'gaussian', 'mitchellcubic'
const INTERPOLATION_KEYS = ['bilinear', 'nearest'];
const INTERPOLATION_METHODS = new Set(INTERPOLATION_KEYS);
/**
 * Preprocessing Resizing Layer
 *
 * This resizes images by a scaling and offset factor
 */
class Resizing extends Layer {
    constructor(args) {
        super(args);
        this.height = args.height;
        this.width = args.width;
        if (args.interpolation) {
            if (INTERPOLATION_METHODS.has(args.interpolation)) {
                this.interpolation = args.interpolation;
            }
            else {
                throw new ValueError(`Invalid interpolation parameter: ${args.interpolation} is not implemented`);
            }
        }
        else {
            this.interpolation = 'bilinear';
        }
        this.cropToAspectRatio = Boolean(args.cropToAspectRatio);
    }
    computeOutputShape(inputShape) {
        inputShape = getExactlyOneShape(inputShape);
        const numChannels = inputShape[2];
        return [this.height, this.width, numChannels];
    }
    getConfig() {
        const config = {
            'height': this.height,
            'width': this.width,
            'interpolation': this.interpolation,
            'cropToAspectRatio': this.cropToAspectRatio
        };
        const baseConfig = super.getConfig();
        Object.assign(config, baseConfig);
        return config;
    }
    call(inputs, kwargs) {
        return tidy(() => {
            const size = [this.height, this.width];
            if (this.interpolation === 'bilinear') {
                return image.resizeBilinear(inputs, size, !this.cropToAspectRatio);
            }
            else if (this.interpolation === 'nearest') {
                return image.resizeNearestNeighbor(inputs, size, !this.cropToAspectRatio);
            }
            else {
                throw new Error(`Interpolation is ${this.interpolation} but only ${[...INTERPOLATION_METHODS]} are supported`);
            }
        });
    }
}
/** @nocollapse */
Resizing.className = 'Resizing';
export { Resizing };
serialization.registerClass(Resizing);
