// @tensorflow/tfjs-layers/dist/layers/nlp/models/backbone.js
/**
 * @license
 * Copyright 2023 Google LLC.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * =============================================================================
 */
/**
 *  Base class for Backbone models.
 */
/* Original source: keras_nlp/models/backbone.py */
import { serialization } from '@tensorflow/tfjs-core';
import { LayersModel } from '../../../engine/training';
import { NotImplementedError } from '../../../errors';
class Backbone extends LayersModel {
    constructor(args) {
        super(args);
    }
    /**
     * A `tf.layers.embedding` instance for embedding token ids.
     */
    get tokenEmbedding() {
        throw new NotImplementedError();
    }
    getConfig() {
        return {
            name: this.name,
            trainable: this.trainable,
        };
    }
    static fromConfig(cls, config) {
        return new cls(config);
    }
}
/** @nocollapse */
Backbone.className = 'Backbone';
export { Backbone };
serialization.registerClass(Backbone);
