// @tensorflow/tfjs-layers/dist/layers/nlp/models/task.js
/**
 * @license
 * Copyright 2023 Google LLC.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * =============================================================================
 */
import { NotImplementedError } from '../../../errors';
import { PipelineModel } from '../utils';
class Task extends PipelineModel {
    constructor(args) {
        super(args);
    }
    checkForLossMismatch(loss) {
        throw new NotImplementedError();
    }
    compile(args) {
        this.checkForLossMismatch(args.loss);
        super.compile(args);
    }
    preprocessSamples(x, y, sampleWeight) {
        throw new NotImplementedError();
    }
    /**
     * A `LayersModel` instance providing the backbone submodel.
     */
    get backbone() {
        return this._backbone;
    }
    set backbone(value) {
        this._backbone = value;
    }
    /**
     * A `LayersModel` instance used to preprocess inputs.
     */
    get preprocessor() {
        return this._preprocessor;
    }
    set preprocessor(value) {
        this.includePreprocessing = value != null;
        this._preprocessor = value;
    }
    getConfig() {
        // Don't chain to super here. The default `getConfig()` for functional
        // models is nested and cannot be passed to our Task constructors.
        throw new NotImplementedError();
    }
    static fromConfig(cls, config) {
        throw new NotImplementedError();
    }
    static backboneCls(cls) {
        return null;
    }
    static preprocessorCls(cls) {
        return null;
    }
    static presets(cls) {
        return {};
    }
    getLayers() {
        throw new NotImplementedError();
    }
    summary(lineLength, positions, printFn = console.log) {
        throw new NotImplementedError();
    }
}
/** @nocollapse */
Task.className = 'Task';
export { Task };
